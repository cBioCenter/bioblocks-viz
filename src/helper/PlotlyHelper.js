"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var data_1 = require("~bioblocks-viz~/data");
/**
 * Generate data in the expected format for a WebGL Scatter plot.
 *
 * @param entry A unit of Plotly data containing points, color, name, and any extras.
 * @param mirrorPoints Should we mirror the points on the x/y axis?
 * @returns Data suitable for consumption by Plotly.
 */
exports.generateScatterGLData = function (entry, mirrorPoints) {
    if (mirrorPoints === void 0) { mirrorPoints = false; }
    return (tslib_1.__assign(tslib_1.__assign({}, exports.generateScatterData(entry, mirrorPoints)), { type: data_1.PLOTLY_CHART_TYPE.scattergl }));
};
/**
 * Generate data in the expected format for a Scatter plot.
 *
 * @param entry A unit of Plotly data containing points, color, name, and any extras.
 * @param mirrorPoints Should we mirror the points on the x/y axis?
 * @returns Data suitable for consumption by Plotly.
 */
exports.generateScatterData = function (entry, mirrorPoints) {
    if (mirrorPoints === void 0) { mirrorPoints = false; }
    var marker = entry.marker, points = entry.points, hoverinfo = entry.hoverinfo, subtitle = entry.subtitle, name = entry.name, text = entry.text;
    var xValues = points.map(function (data) { return data.i; });
    var yValues = points.map(function (data) { return data.j; });
    var zValues = points.map(function (data) { return (data.dist ? data.dist : -1); });
    var textValues = text ? (Array.isArray(text) ? text : [text]) : [];
    return {
        hoverinfo: hoverinfo ? hoverinfo : 'x+y+z',
        marker: tslib_1.__assign({ color: derivePlotlyColor(mirrorPoints, zValues, entry), size: entry.nodeSize }, marker),
        mode: 'markers',
        name: name + " " + subtitle,
        text: deriveScatterText(mirrorPoints, textValues),
        type: data_1.PLOTLY_CHART_TYPE.scatter,
        x: derivePoints(mirrorPoints, xValues, yValues),
        y: derivePoints(mirrorPoints, yValues, xValues),
        z: derivePoints(mirrorPoints, zValues, zValues),
    };
};
var derivePlotlyColor = function (mirrorPoints, zValues, entry) {
    var totalColors = mirrorPoints ? zValues.length * 2 : zValues.length;
    var result = new Array(totalColors);
    if (entry.marker && typeof entry.marker.color === 'string') {
        return result.fill(entry.marker.color);
    }
    else {
        var zStrings = zValues.map(function (val) { return val.toString(); });
        return mirrorPoints ? tslib_1.__spread(zStrings, zStrings) : zStrings;
    }
};
var derivePoints = function (mirrorPoints, points, oppositePoints) {
    return mirrorPoints ? tslib_1.__spread(points, oppositePoints) : points;
};
var deriveScatterText = function (mirrorPoints, textValues) {
    return mirrorPoints
        ? tslib_1.__spread(textValues, textValues.map(
        // Given a coordinate '(x, y)', create '(y, x)' - needed because we have custom hover labels.
        function (coord) {
            if (coord.includes('<br>') && coord.includes('(')) {
                var breakIndex = coord.indexOf('<br>');
                return "(" + coord
                    .substr(1, breakIndex - 2)
                    .split(', ')
                    .reverse()
                    .join(', ') + ")" + coord.substring(breakIndex);
            }
            else if (coord.includes('(')) {
                return "(" + coord
                    .substr(1, coord.length - 2)
                    .split(', ')
                    .reverse()
                    .join(', ') + ")";
            }
            else {
                return "" + coord
                    .split(', ')
                    .reverse()
                    .join(', ');
            }
        })) : textValues;
};
/**
 * Determines the appropriate hover text in plotly for this coupling score.
 *
 * Currently the following 3 fields will be appended if present:
 * - Amino acid (A_i, A_j)
 * - Score
 * - Probability
 */
exports.generateCouplingScoreHoverText = function (point) {
    var hoverText = '';
    if (point) {
        // Yields the format '(2B, 9S)' if amino acid is known, otherwise just '(2, 9)'
        hoverText =
            point.A_i && point.A_j ? "(" + point.i + point.A_i + ", " + point.j + point.A_j + ")" : "(" + point.i + ", " + point.j + ")";
        if (point.score) {
            hoverText = hoverText + "<br>Score: " + point.score;
        }
        if (point.probability) {
            hoverText = hoverText + "<br>Probability: " + point.probability.toFixed(1);
        }
    }
    return hoverText;
};
exports.generateFloat32ArrayFromContacts = function (array) {
    var result = new Float32Array(array.length * 2);
    array.forEach(function (item, index) {
        result[index * 2] = item.i;
        result[index * 2 + 1] = item.j;
    });
    return result;
};
/**
 * Generate data in the expected format for a Plotly PointCloud.
 *
 * @param entry A unit of Plotly data containing points, color, and any extras.
 * @param mirrorPoints Should we mirror the points on the x/y axis?
 * @returns Data suitable for consumption by Plotly.
 */
exports.generatePointCloudData = function (entry, mirrorPoints) {
    if (mirrorPoints === void 0) { mirrorPoints = false; }
    var points = entry.points;
    var coords = exports.generateFloat32ArrayFromContacts(points);
    return {
        marker: tslib_1.__assign(tslib_1.__assign({}, entry.marker), { sizemax: entry.nodeSize * 2, sizemin: entry.nodeSize }),
        mode: 'markers',
        type: data_1.PLOTLY_CHART_TYPE.pointcloud,
        xy: mirrorPoints
            ? new Float32Array(tslib_1.__spread(Array.from(coords), Array.from(coords)
                .slice()
                .reverse()))
            : coords,
    };
};
//# sourceMappingURL=PlotlyHelper.js.map