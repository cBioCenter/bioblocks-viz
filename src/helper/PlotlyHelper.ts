import { IContactMapChartData } from '~bioblocks-viz~/component';
import { ICouplingScore, IPlotlyData, PLOTLY_CHART_TYPE } from '~bioblocks-viz~/data';

/**
 * Generate data in the expected format for a WebGL Scatter plot.
 *
 * @param entry A unit of Plotly data containing points, color, name, and any extras.
 * @param mirrorPoints Should we mirror the points on the x/y axis?
 * @returns Data suitable for consumption by Plotly.
 */
export const generateScatterGLData = (
  entry: IContactMapChartData,
  mirrorPoints: boolean = false,
): Partial<IPlotlyData> => ({
  ...generateScatterData(entry, mirrorPoints),
  type: PLOTLY_CHART_TYPE.scattergl,
});

/**
 * Generate data in the expected format for a Scatter plot.
 *
 * @param entry A unit of Plotly data containing points, color, name, and any extras.
 * @param mirrorPoints Should we mirror the points on the x/y axis?
 * @returns Data suitable for consumption by Plotly.
 */
export const generateScatterData = (
  entry: IContactMapChartData,
  mirrorPoints: boolean = false,
): Partial<IPlotlyData> => {
  const { marker, points, hoverinfo, subtitle, name, text } = entry;
  const xValues = points.map(data => data.i);
  const yValues = points.map(data => data.j);
  const zValues = points.map(data => (data.dist ? data.dist : -1));
  const textValues = text ? (Array.isArray(text) ? text : [text]) : [];

  return {
    hoverinfo: hoverinfo ? hoverinfo : 'x+y+z',
    marker: {
      color: derivePlotlyColor(mirrorPoints, zValues, entry),
      size: entry.nodeSize,
      ...marker,
    },
    mode: 'markers',
    name: `${name} ${subtitle}`,
    text: deriveScatterText(mirrorPoints, textValues),
    type: PLOTLY_CHART_TYPE.scatter,
    x: derivePoints(mirrorPoints, xValues, yValues),
    y: derivePoints(mirrorPoints, yValues, xValues),
    z: derivePoints(mirrorPoints, zValues, zValues),
  };
};

const derivePlotlyColor = (mirrorPoints: boolean, zValues: number[], entry: IContactMapChartData) => {
  const totalColors = mirrorPoints ? zValues.length * 2 : zValues.length;
  const result = new Array<Plotly.Color>(totalColors);

  if (entry.marker && typeof entry.marker.color === 'string') {
    return result.fill(entry.marker.color);
  } else {
    const zStrings = zValues.map(val => val.toString());

    return mirrorPoints ? [...zStrings, ...zStrings] : zStrings;
  }
};

const derivePoints = (mirrorPoints: boolean, points: number[], oppositePoints?: number[]) =>
  mirrorPoints ? [...points, ...oppositePoints] : points;

const deriveScatterText = (mirrorPoints: boolean, textValues: string[]) => {
  return mirrorPoints
    ? [
        ...textValues,
        ...textValues.map(
          // Given a coordinate '(x, y)', create '(y, x)' - needed because we have custom hover labels.
          coord => {
            if (coord.includes('<br>') && coord.includes('(')) {
              const breakIndex = coord.indexOf('<br>');

              return `(${coord
                .substr(1, breakIndex - 2)
                .split(', ')
                .reverse()
                .join(', ')})${coord.substring(breakIndex)}`;
            } else if (coord.includes('(')) {
              return `(${coord
                .substr(1, coord.length - 2)
                .split(', ')
                .reverse()
                .join(', ')})`;
            } else {
              return `${coord
                .split(', ')
                .reverse()
                .join(', ')}`;
            }
          },
        ),
      ]
    : textValues;
};
/**
 * Determines the appropriate hover text in plotly for this coupling score.
 *
 * Currently the following 3 fields will be appended if present:
 * - Amino acid (A_i, A_j)
 * - Score
 * - Probability
 */
export const generateCouplingScoreHoverText = (point: ICouplingScore) => {
  let hoverText = '';
  if (point) {
    // Yields the format '(2B, 9S)' if amino acid is known, otherwise just '(2, 9)'
    hoverText =
      point.A_i && point.A_j ? `(${point.i}${point.A_i}, ${point.j}${point.A_j})` : `(${point.i}, ${point.j})`;

    if (point.score) {
      hoverText = `${hoverText}<br>Score: ${point.score}`;
    }
    if (point.probability) {
      hoverText = `${hoverText}<br>Probability: ${point.probability.toFixed(1)}`;
    }
  }

  return hoverText;
};

export const generateFloat32ArrayFromContacts = (array: Array<{ i: number; j: number }>) => {
  const result = new Float32Array(array.length * 2);
  array.forEach((item, index) => {
    result[index * 2] = item.i;
    result[index * 2 + 1] = item.j;
  });

  return result;
};

/**
 * Generate data in the expected format for a Plotly PointCloud.
 *
 * @param entry A unit of Plotly data containing points, color, and any extras.
 * @param mirrorPoints Should we mirror the points on the x/y axis?
 * @returns Data suitable for consumption by Plotly.
 */
export const generatePointCloudData = (
  entry: IContactMapChartData,
  mirrorPoints: boolean = false,
): Partial<IPlotlyData> => {
  const { points } = entry;
  const coords = generateFloat32ArrayFromContacts(points);

  return {
    marker: {
      ...entry.marker,
      sizemax: entry.nodeSize * 2,
      sizemin: entry.nodeSize,
    },
    mode: 'markers',
    type: PLOTLY_CHART_TYPE.pointcloud,
    xy: mirrorPoints
      ? new Float32Array([
          ...Array.from(coords),
          ...Array.from(coords)
            .slice()
            .reverse(),
        ])
      : coords,
  };
};
