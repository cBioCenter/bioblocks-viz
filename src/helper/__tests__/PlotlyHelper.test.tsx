import { generatePointCloudData, generateScatterGLData } from '../PlotlyHelper';

describe('generateScatterGLData', () => {
  const sampleInput = {
    marker: {
      color: 'blue',
    },
    name: 'da-bo-dee',
    nodeSize: 4,
    points: [
      {
        dist: 6,
        i: 9,
        j: 8,
      },
      { dist: 2, i: 7, j: 6 },
    ],
  };

  it('Should create the expected plotly scatter data format when given defaults.', () => {
    const result = generateScatterGLData(sampleInput, false);
    expect(result).toMatchSnapshot();
  });

  it('Should create the expected plotly scatter data format when points are mirrored.', () => {
    const result = generateScatterGLData(sampleInput, true);
    expect(result.x).toEqual([9, 7, 8, 6]);
    expect(result.y).toEqual([8, 6, 9, 7]);
  });

  it('Should use z values for color when points are not mirrored and no explicit color is provided.', () => {
    const result = generateScatterGLData({ ...sampleInput, marker: {} }, false);
    expect(result.marker && result.marker.color).toEqual(['6', '2']);
  });
});

describe('generatePointCloudData', () => {
  const sampleInput = {
    color: 'blue',
    name: 'da-bo-dai',
    nodeSize: 5,
    points: [
      {
        dist: 4,
        i: 1,
        j: 2,
      },
      { dist: 5, i: 3, j: 4 },
    ],
  };

  it('Should create the expected plotly point cloud data format when given defaults.', () => {
    const result = generatePointCloudData(sampleInput, false);
    expect(result).toMatchSnapshot();
  });

  it('Should create the expected plotly point cloud data format when points are mirrored.', () => {
    const result = generatePointCloudData(sampleInput, true);
    const expected = new Float32Array([1, 2, 3, 4, 4, 3, 2, 1]);
    expect(result.xy).toEqual(expected);
  });
});
