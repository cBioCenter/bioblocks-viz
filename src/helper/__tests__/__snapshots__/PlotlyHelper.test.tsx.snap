// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generatePointCloudData Should create the expected plotly point cloud data format when given defaults. 1`] = `
Object {
  "marker": Object {
    "sizemax": 10,
    "sizemin": 5,
  },
  "mode": "markers",
  "type": "pointcloud",
  "xy": Float32Array [
    1,
    2,
    3,
    4,
  ],
}
`;

exports[`generateScatterGLData Should create the expected plotly scatter data format when given defaults. 1`] = `
Object {
  "hoverinfo": "x+y+z",
  "marker": Object {
    "color": "blue",
    "size": 4,
  },
  "mode": "markers",
  "name": "da-bo-dee undefined",
  "text": Array [],
  "type": "scattergl",
  "x": Array [
    9,
    7,
  ],
  "y": Array [
    8,
    6,
  ],
  "z": Array [
    6,
    2,
  ],
}
`;
